api/src/main/java/org/apache/iceberg/io/FileIOMetricsContext.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import org.apache.iceberg.metrics.MetricsContext;
+
+/**
+ * Extension of MetricsContext for use with FileIO to define standard metrics
+ * that should be reported.
+ */
+public interface FileIOMetricsContext extends MetricsContext {
```

Comment:
I like having a place for the constants, but what is the value of making an identical subinterface of `MetricsContext` rather than using the parent directly?

Also, instead of just names, should we have some metric definition that includes unit (assuming we move forward with units)?

Reply:
I think the intention is to have `HadoopFIleIOMetricsContext` accessing these, but we are still switching across different metrics in the implementation and we do not allow unsupported ones. In that case, does it make more sense to just have methods for updating each metrics directly?

Reply:
Yes, these are metrics that are specific to file system operations to distinguish from other operations.  We could collapse them all, but it could get confusing and maybe we don't want to reuse the same names for different usages.

The unit should be defined by the producer, not required by the metric.  For example, one file IO impl could produce bytes while the other could produce kilobytes.  As long as they annotate the unit they're producing correctly, it should be fine.

Reply:
I understand coordinating what metrics are required, but why should `FileIO` require a specific `MetricsContext` sub-interface? Why not be able to supply any `MetricsContext` with these metric names?

Reply:
`FileIO` does not specifically require a sub-interface.  You can see that S3FileIO only as a handle on a `MetricsContext`.  The only reason for the sub-interface is to have a place to put FileIO specific things, though right now that is only the name of the generic metric names.  These metrics need to be translated by the metrics implementations (in the hadoop case it would be `read.bytes -> statistics.incrementBytesRead()`).

We cannot put the metric names in the HadoopMetricsContext because that would pull in dependencies that would fail class loading when hadoop is not available. 

Alternatively we could just have a an interface/class as the holder for FileIO Metrics, but since this is scoped to usage within the FileIO space, it makes sense to just define a sub-interface to expose the metrics names.

Reply:
Other implementations of metrics (for example if there is a specific Flink metrics framework) would need to handle these same metric names since that is what the FileIO implementations are producing.

The structure of the interfaces is just to help keep that scoped to FileIO and what is expected to be produce by FileIO implementations.

Reply:
I think it's a bit confusing to create this interface just for metric names. I'd probably add the metric names as `METRIC_READ_BYTES` and similar on the FileIO interface itself. But this is fairly minor as long as we aren't actually requiring this interface anywhere in code.